mtags/src/main/scala-3/scala/meta/internal/pc/SignatureHelpProvider.scala

```diff
@@ -0,0 +1,162 @@
+import dotty.tools.dotc.ast.tpd.Tree
+import dotty.tools.dotc.core.Constants.Constant
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Denotations.SingleDenotation
+import dotty.tools.dotc.core.Symbols
+import dotty.tools.dotc.core.Symbols.*
+import dotty.tools.dotc.core.Types.ErrorType
+import dotty.tools.dotc.interactive.Interactive
+import dotty.tools.dotc.interactive.InteractiveDriver
+import dotty.tools.dotc.reporting.AmbiguousOverload
+import dotty.tools.dotc.reporting.NoMatchingOverload
+import dotty.tools.dotc.util.Signatures
+import dotty.tools.dotc.util.SourcePosition
+import dotty.tools.dotc.util.Spans.Span
+import org.eclipse.lsp4j.ParameterInformation
+import org.eclipse.lsp4j.SignatureHelp
+import org.eclipse.lsp4j.SignatureInformation
+import dotty.tools.dotc.core.Types.PolyType
+import dotty.tools.dotc.core.Types.Type
+import dotty.tools.dotc.core.Types.MethodType
+import scala.annotation.tailrec
+import scala.collection.immutable.LazyList.cons
+import dotty.tools.dotc.core.Flags
+
+object SignatureHelpProvider:
+
+  def contribute(driver: InteractiveDriver, params: OffsetParams) =
+    val uri = params.uri
+    val sourceFile = CompilerInterfaces.toSource(params.uri, params.text)
+    driver.run(uri, sourceFile)
+
+    given ctx: Context = driver.currentCtx
+    val pos = driver.sourcePosition(params)
+    val trees = driver.openedTrees(uri)
+    val path =
+      Interactive
+        .pathTo(trees, pos)(using ctx)
+        .dropWhile {
+          case _: tpd.Apply => false
+          case unapply: tpd.UnApply =>
+            // Don't show tuple unapply method
+            val sym = unapply.fun.symbol.owner.companionClass
+            Symbols.defn.isTupleClass(sym)
+          case _ => true
+        }
+
+    val (paramN, callableN, alternatives) =
+      callInfo(path, pos.span)(using ctx)
+
+    val signatureInfos =
+      alternatives.flatMap(Signatures.toSignature(_)(using ctx))
+    new SignatureHelp(
+      signatureInfos.map(signatureToSignatureInformation).asJava,
+      callableN,
+      paramN
+    )
+  end contribute
+
+  private def callInfo(path: List[Tree], span: Span)(using
+      ctx: Context
+  ): (Int, Int, List[SingleDenotation]) =
+    Signatures.callInfo(path, span)(using ctx) match
+      case default @ (_, _, Nil) =>
+        path match
+          case (tpd.UnApply(fun, _, patterns)) :: parent :: _ =>
+            val implicitsBefore = countParams(fun)
+
+            def defaultParamIndex =
+              patterns.indexWhere(_.span.contains(span)) match
+                case -1 => (patterns.length - 1 max 0) + implicitsBefore
+                case n => n + implicitsBefore
+
+            val funSymbol = fun.symbol
+            val retType = funSymbol.info.finalResultType
+            val isSomeMatch = funSymbol.owner.companionClass == defn.SomeClass
+            val isExplicitUnapply =
+              !isSomeMatch && retType.typeSymbol == defn.OptionClass
```

Comment:
>isExplicitUnapply will be true when we explicitly call unapply method like Foo.unapply(@@)

It might be, but we will be at an Apply node, so the code will not be invoked.

Reply:
Auch, I got it! `isExplicitUnapply = true` means `unapply` method is **explicitly defined**, not **explicitly called**